docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,429 @@
+require that all existing data is validated against the proposed schema
+change before completion. This would include conversions such as
+`STRING(100) -> STRING(10)` or `BYTES -> STRING`. These have a cost
+somewhere between trivial and general conversions, since on-disk data
+does not need to be rewritten, only revalidated.
+
+### Cast
+
+Cast conversions involve applying a built-in cast behavior to the
+existing column data requiring the on-disk KV data to be rewritten, such
+as `UUID -> STRING`.
+
+### Assisted
+
+Assisted conversions involve applying a `USING expression` to the
+existing column data, such as `FLOAT -> INT USING ceil(column_name)`.
+
+## Indexed columns
+
+Columns that are part of an index are supported, however columns that
+are part of a primary key are explicitly out of scope for this change.
+
+## Two-phase change
+
+Given that a type-change process may take an arbitrary and unpredictable
+amount of time to complete, we support an optional two-phase approach to
+making the schema change visible to SQL clients. This allows the user to
+choose exactly when the change to the table schema will become visible,
+allowing application downtime to be minimized in non-trivial-change
+cases.
+
+The `ALTER TABLE` command is first run with the `DEFERRED` option, which
+begins the type-conversion in the background. At any point in time, but
+generally after the job has completed, the user may issue the same
+`ALTER TABLE` command, replacing `DEFERRED` with `COMPLETE` to allow the
+type change to become visible once the job has completed. The `ABORT`
+option allows an ongoing type-conversion operation to be canceled.
+
+```
+-- Convert a STRING to the 2.0 INET type.
+-- This command will return a row containing a job id.
+ALTER TABLE rpc_logs
+  ALTER COLUMN caller_ip
+  SET DATA TYPE INET
+  DEFERRED
+
+-- The application can continue reading and writing this column as STRING.
+
+-- Wait for the job to finish, stop the application, swap the column type.
+-- This command would block, returning any output from the conversion job.
+-- Any reads that happen-after this command will see the new type, while any
+-- reads that happen-before this command will see the old type.
+ALTER TABLE rpc_logs
+  ALTER COLUMN caller_ip
+  SET DATA TYPE INET
+  COMPLETE
+
+-- Restart new version of the application that expects the INET type.
+```
+
+Electing not to use the `DEFERRED` option is equivalent to running a
+`DEFERRED` followed immediately by a `COMPLETE`. We will refer to this
+as "auto-complete" mode.
+
+The `COMPLETE` option requires the entire, original `ALTER TABLE`
+command as a sanity-check. If the type conversion generated any output,
+it will be emitted by this command. If the type conversion has not yet
+completed, this command will block until it has. It is not an error to
+run `COMPLETE` before the conversion job has finished; it is simply a
+signal to indicate that the SQL consumers are prepared to handle the new
+column type.
+
```

Comment:
good thinking about this. I'm not sure if we need to do it in a first implementation though.